### DISCUSSION TITLE
Getting looser throw specifier error when trying to mock a class which uses kj Promises

### DISCUSSION THREAD

**Discussion by legolas145** [google/googletest#4511] [Community Help]
Hello,
I am pretty new to Gtest and Gmock.I have a interface which has a function that returns a kj::Promise as response.I am using mock to mock that interface.

```
//suggestion_interface.h
#pragma once

#include <string>
#include <kj/async.h>

class ISuggestions {
 public:

  // Function to get the suggestions results from swype for a give text input
  virtual kj::Promise<Suggestions> GetAutocompleteSuggestions(std::string input) = 0;

};



//suggestion_mock.h
#include <gmock/gmock.h>

class MockSuggestions : public ISuggestions {
 public:

  MOCK_METHOD(kj::Promise<Suggestions>, GetAutocompleteSuggestions,(std::string input),(override));

};

```

I notice that I am getting a looser throw specifier error
/usr/include/gmock/gmock-spec-builders.h: In instantiation of 'class testing::internal::ActionResultHolder<kj::Promise<Suggestions> >':
/usr/include/gmock/gmock-spec-builders.h:1601:20:   required from 'testing::internal::FunctionMocker<R(Args ...)>::Result testing::internal::FunctionMocker<R(Args ...)>::Invoke(Args ...) [with R = kj::Promise<Suggestions>; Args = {std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >}; testing::internal::FunctionMocker<R(Args ...)>::Result = kj::Promise<Suggestions>]'

/usr/include/gmock/gmock-spec-builders.h:1378:7: error: looser throw specifier for 'virtual testing::internal::ActionResultHolder<kj::Promise<Suggestions> >::~ActionResultHolder() noexcept (false)'
 1378 | class ActionResultHolder : public UntypedActionResultHolderBase {
      |       ^~~~~~~~~~~~~~~~~~
/usr/include/gmock/gmock-spec-builders.h:1370:11: note: overridden function is 'virtual testing::internal::UntypedActionResultHolderBase::~UntypedActionResultHolderBase() noexcept'

May I know why I am getting this error and what should I do to fix this?

**Comment by Owen836352o3** [comment]
Hi

**Reply by legolas145** [reply]
Hey @Owen836352o3 . Have you also encountered this issue before? Let me know if you have any resolutions for this

**Comment by Shriharsh-Deshmukh** [comment]
The "looser throw specifier" error occurs because `kj::Promise` has a non-`noexcept` destructor, but Google Mock (GMock) expects that all mocked method return types have `noexcept` destructors.

### Solution:
1. **Wrap `kj::Promise` in a custom type**: Create a wrapper around `kj::Promise` that has a `noexcept` destructor.

2. **Example**:
   ```cpp
   class NoexceptPromiseWrapper {
   public:
       kj::Promise<Suggestions> promise;
       ~NoexceptPromiseWrapper() noexcept = default;
   };

   class MockSuggestions : public ISuggestions {
   public:
       MOCK_METHOD(NoexceptPromiseWrapper, GetAutocompleteSuggestions, (std::string input), (override));
   };
   ```

   This ensures that the `noexcept` constraint is satisfied and should resolve the error.

This approach keeps `kj::Promise` usage while aligning with GMock's `noexcept` requirements.
